src/lib/io/sol-i2c-common.c

```diff
@@ -23,20 +23,259 @@
+    uint8_t i;
+
+    SOL_VECTOR_FOREACH_IDX (&i2c_shared_vector, tmp, i) {
+        if (tmp->i2c == i2c) {
+            shared = tmp;
+            break;
+        }
+    }
+    SOL_NULL_CHECK_MSG(shared, NULL, "Internal I2C bus handle not found.");
+
+    op_set = calloc(1, sizeof(struct sol_i2c_op_set));
+    SOL_NULL_CHECK_MSG(op_set, NULL, "Could not allocate memory for the I2C Operation Set.");
+
+    op_set->set = set;
+    op_set->cb = cb;
+    op_set->cb_data = cb_data;
+    op_set->delay = delay;
+    op_set->addr = addr;
+
+    if (sol_ptr_vector_append(&shared->dispatcher.queue, op_set)) {
+        SOL_ERR("Couldn't append operation set on dispatcher queue.");
+        goto error;
+    }
+
+    if (sol_ptr_vector_get_len(&shared->dispatcher.queue) == 1) {
+        // Don't check on timer to verify that the dispatcher engine needs to be started,
+        // it can be between exec -> done operation and this would set a timer with incorrect
+        // timeout value for operations set that uses specific time in between.
+
+        // This should always be null.
+        SOL_NULL_CHECK_MSG_GOTO(!shared->dispatcher.timer, start_error,
+            "Dispatcher timer should always be NULL in this point.");
+
+        shared->dispatcher.timer = sol_timeout_add(0, i2c_dispatcher_exec_op, shared);
+        SOL_NULL_CHECK_MSG_GOTO(shared->dispatcher.timer, start_error,
+            "Couldn't start I2C Dispatcher.");
+    }
+
+    return (struct sol_i2c_op_set_pending *)op_set;
+
+start_error:
+    sol_ptr_vector_del(&shared->dispatcher.queue, 0);
+error:
+    free(op_set);
+    return NULL;
+}
+
+SOL_API void
+sol_i2c_dispatcher_remove_op_set(struct sol_i2c *i2c, struct sol_i2c_op_set_pending *pending)
+{
+    struct sol_i2c_shared *tmp, *shared = NULL;
+    struct sol_i2c_dispatcher *dispatcher = NULL;
+    struct sol_ptr_vector *pv = NULL;
+    struct sol_i2c_op_set *op_set = NULL;
+    void *t;
+    uint8_t i;
+
+    SOL_VECTOR_FOREACH_IDX (&i2c_shared_vector, tmp, i) {
+        if (tmp->i2c == i2c) {
+            shared = tmp;
+            break;
+        }
+    }
+    if (!shared)
+        return;
+
+    pv = &shared->dispatcher.queue;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (pv, t, i) {
+        if (t == pending) {
+            op_set = t;
+            break;
+        }
+    }
+    if (!op_set)
+        return;
+
+    if (!i) {
+        // current set is in execution, delay the deletion until current operation finishes
+        op_set->delete_me = true;
+    } else {
+        sol_ptr_vector_del(pv, i);
+        free(op_set);
+    }
+}
+
+void
+i2c_dispatcher_close(struct sol_i2c *i2c, struct sol_i2c_dispatcher *dispatcher)
+{
+    struct sol_i2c_op_set *op_set;
+    uint8_t i;
+
+    if (dispatcher->timer)
+        sol_timeout_del(dispatcher->timer);
+
+    if (dispatcher->pending)
+        sol_i2c_pending_cancel(i2c, dispatcher->pending);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&dispatcher->queue, op_set, i) {
+        free(op_set);
```

Comment:
Is owned by the user and can be static some times.

On Mon, May 23, 2016 at 11:12 AM, Iván Briano <notifications@github.com>
wrote:

> In src/lib/io/sol-i2c-common.c
> <https://github.com/solettaproject/soletta/pull/2051#discussion_r64229126>
> :
>
> > +}
> > +
> > +void
> > +i2c_dispatcher_close(struct sol_i2c *i2c, struct sol_i2c_dispatcher *dispatcher)
> > +{
> > +    struct sol_i2c_op_set *op_set;
> > +    uint8_t i;
> > +
> > +    if (dispatcher->timer)
> > +        sol_timeout_del(dispatcher->timer);
> > +
> > +    if (dispatcher->pending)
> > +        sol_i2c_pending_cancel(i2c, dispatcher->pending);
> > +
> > +    SOL_PTR_VECTOR_FOREACH_IDX (&dispatcher->queue, op_set, i) {
> > +        free(op_set);
>
> What about the vector inside the set?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/solettaproject/soletta/pull/2051/files/adc0c6ac58fbaffd48d71795f0c219d2d19505bb#r64229126>
>